SWTCH-1197 creating suborganization in suborganization view

dawidgil opened the pull request:
Fix a regression with not working button in sub-organization view.

jrhender commented on src/app/routes/applications/governance-list/governance-list.component.ts:
```diff
@@ -368,7 +370,27 @@ export class GovernanceListComponent implements OnInit, OnDestroy {
+      .subscribe(async () => await this.newSubOrg(parentOrg));
+  }
+
   async newSubOrg(parentOrg: any) {
```
I think that `newSubOrg` and `createSubOrg` are a bit too similar a name.  How about this rename?
```suggestion
  async viewNewSubOrg(parentOrg: any) {
```

dawidgil replied:
Ok, sure i can rename it. I just wanted to remove regression before deploying to volta and handle it in next few days properly.

jrhender commented on src/app/routes/applications/governance-list/governance-list.component.ts:
```diff
@@ -368,7 +370,27 @@ export class GovernanceListComponent implements OnInit, OnDestroy {
+      .subscribe(async () => await this.newSubOrg(parentOrg));
+  }
+
   async newSubOrg(parentOrg: any) {
```
If this is only being called in component, can it be made private?
```suggestion
  private async newSubOrg(parentOrg: any) {
```

dawidgil replied:
This is used in template as well.
I refactored this once and that's why this regression happened. I didn't expected that parent component is using child component methods.
I created there components for actions, and those components are handling displaying dialogs etc. method like `newSubOrg` is used in template just to refresh the list after closing a dialog.

jrhender replied:
Ok, fair enough 👍

jrhender commented on src/app/routes/applications/governance-list/governance-list.component.ts:
```diff
@@ -368,7 +370,27 @@ export class GovernanceListComponent implements OnInit, OnDestroy {
+      disableClose: true
+    });
+
+    dialogRef.afterClosed()
+      .pipe(
+        filter(Boolean),
+        takeUntil(this.subscription$))
+      .subscribe(async () => await this.newSubOrg(parentOrg));
```
This is probably due to my lack of RxJs knowledge but could you help me understand what this code is doing?

- In `pipe` you only take the not defined/not-null values (due to the `filter(Boolean)`). What happens if the return from `afterClosed()` is `undefined`?
- In `pipe`, `takeUntil(this.subscription$)` is a little unclear. Like I get that it means take until `this.subscription$` emits, but why should we do this? What is `this.subscription$` observing? Is this an Angular convention?
- `this.newSubOrg` is called on the `subscribe`. Does this mean that in theory `this.newSubOrg` could be called more than once?

dawidgil replied:
Ah now i remember. Filtering is needed because we can send there some values while closing dialogs. We are sending true/false values there. True when we want to refresh data and false if just closing a dialog.

jrhender replied:
Ok, thanks for the explanations

jrhender commented on src/app/routes/applications/governance-list/governance-list.component.ts:
```diff
@@ -368,7 +370,27 @@ export class GovernanceListComponent implements OnInit, OnDestroy {
+
+    dialogRef.afterClosed()
+      .pipe(
+        filter(Boolean),
```
I was reading a bit about `filter(Boolean)` and found this SO answer which suggests using a explicit filter for `undefined` so that the type is preserved. Have you seen this before? https://stackoverflow.com/a/57928681
```suggestion
        filter(v=> v!== undefined),
```

jrhender replied:
Ah, maybe it's better to just provide the type like `filter<type>(Boolean)` but maybe that's not necessary in this case?

dawidgil replied:
hmm, interesting. Thanks for sharing this one.

jrhender reviewed (commented):
A few questions/suggestions. Probably I'm a bit to unfamiliar with Angular/RxJs 😓

dawidgil commented:
@jrhender I refactored a lot here and fixed few bugs/inconsistent behavior. I used ngrx to handle most of things in organization list view. I also created a new component for handling only organization list. I want to do the same with applications and roles in future and extract common stuff to abstract class. I also added tests to most of things.

jrhender commented on src/app/routes/applications/organization-list/organization-list.component.html:
```diff
@@ -0,0 +1,114 @@
+<!-- Filter Sub Organization -->
```
Does it make sense to you that this new `organization-list.component` is under `routes/applications`. What does "applications" mean in this context 🤔 ?

dawidgil replied:
No, it doesn't. I forgot to change application.component.ts into governance.component.ts. I should change also routing here. Thanks for pointing this out!

jrhender commented on src/app/state/organization/utils/get-main-orgs.ts:
```diff
@@ -0,0 +1,40 @@
+  if (list.length || subList.length) {
+    for (let i = 0; i < subList.length; i++) {
+      const arr = subList[i];
+      if (arr && arr.length) {
```
The subsequent code doesn't seem to be referencing the `length` property of `arr`? Why the check for it in this `if`?

dawidgil replied:
Ok, this code is just extracted from governance-list.component.ts :D I didn't touched that. I just wrote few tests to be sure that it is working correctly. I will refactor this.
To let you know what is going on here, this code is getting main organization (for example test.iam.ewc) and all it's direct suborganization (suborg.test.iam.ewc etc.), but not sub-(...)-sub-org.

jrhender commented on src/app/state/organization/utils/get-main-orgs.ts:
```diff
@@ -0,0 +1,40 @@
+  const list = [];
+  const subList = [];
+
+  // Separate Parent & Child Orgs
+  $getOrgList.forEach((item: any) => {
+    const namespaceArr = item.namespace.split('.');
+    if (namespaceArr.length === 3) {
+      list.push(item);
+    } else {
+      if (!subList[namespaceArr.length - 4]) {
+        subList[namespaceArr.length - 4] = [];
+      }
+      subList[namespaceArr.length - 4].push(item);
+    }
+  });
```
The `3` and `4` are a bit of a magic numbers. I think constants should be used.
I'm not sure why the `4`. If you had a suborg of `finance.dawidcorp.iam.ewc`, `namespaceArr.length` would be `4`. So I guess what you're trying to do is get lists of the suborg that are at each "level"?
```suggestion
  // Separate Parent & Child Orgs
   const numberOfLabelsInOrgNamespace = 3;
  $getOrgList.forEach((item: any) => {
    const namespaceArr = item.namespace.split('.');
    if (namespaceArr.length === numberOfLabelsInOrgNamespace) {
      list.push(item);
    } else {
      if (!subList[namespaceArr.length - 4]) {
        subList[namespaceArr.length - 4] = [];
      }
      subList[namespaceArr.length - 4].push(item);
    }
  });
```

jrhender commented on src/app/state/organization/utils/get-main-orgs.ts:
```diff
@@ -0,0 +1,40 @@
+export const getMainOrgs = ($getOrgList: any[]) => {
```
I find this function pretty confusing, maybe it could use some documentation

jrhender commented on src/app/state/organization/utils/get-main-orgs.ts:
```diff
@@ -0,0 +1,40 @@
+      if (arr && arr.length) {
+        for (const subOrg of arr) {
+          let exists = false;
+          for (const mainOrg of list) {
+            if (subOrg.namespace && subOrg.namespace.includes(mainOrg.namespace)) {
+              exists = true;
+              break;
+            }
+          }
+          if (!exists) {
+            list.push(subOrg);
+          }
```
It seems like you are trying to figure out if a `subOrg`'s parent Org is in the `list`? Is it possible to have different result based on the order of the elements in `arr`? Is you're checking the elements of `list` but also adding to `list` in each iteration?

jrhender reviewed (commented):
Looks good, I think the `get-main-orgs` function could just use some additional clarification

dawidgil commented on src/app/routes/applications/governance-list/governance-list.component.html:
```diff
@@ -138,18 +97,6 @@
               <span>View Details</span>
             </button>
 
-            <app-organization-actions *ngIf="listType === ListType.ORG"
-                                      [organization]="element"
-                                      (organizationCreated)="newSubOrg(element)"
-                                      (viewApps)="viewApps(ListType.ORG, element)"
-                                      (viewRoles)="viewRoles(ListType.ORG, element)"
-                                      (appCreated)="newApp(element)"
-                                      (roleCreated)="viewRoles(ListType.ORG, element)"
-                                      (edited)="edit()"
-                                      (transferred)="transferOwnership()"
-                                      (deleteConfirmed)="remove(listType, element)">
-            </app-organization-actions>
-
```
No governance-list is for application and roles right now. organization-list is only for organization. I didn't want to refactor everything, but the plan is to split application and roles into separate components.

jrhender reviewed (approved):
👍
